index.d.ts

```diff
@@ -43,6 +44,16 @@ declare interface CPUMetric {
   raw: number
 }
 
+declare interface ResourceUsageMetric {
+  /**
+    * Cpu usage percentage
+    */
+   cpu: number,
+   /**
+     * Raw vaule returned by `process.resourceUsage()`
+     */
+   raw: NodeJS.ResourceUsage
```

Comment:
I want to try to use Node built-in types, are there any drawbacks in doing so @acheronfail ? For example, is this portable across different Node versions? Does the fact the `@types/node` is a dev dependency cause any problems to the user (those dependencies are not installed when you do `npm i`). I apologize if these are dumb questions.

Reply:
For context #109

Reply:
> I apologize if these are dumb questions.

There is no such thing!

> Does the fact the @types/node is a dev dependency cause any problems to the user

No, users of the library that have TypeScript must already have the `@types/node` package installed for the Node version that they're using anyway.

> I want to try to use Node built-in types, are there any drawbacks in doing so

The only drawback would be if a user was on a previous version of Node (and therefore `@types/node`) that didn't include the type that you were using. But, this isn't _really_ an issue, because if they're on an older version of Node that this package doesn't support, then it won't work anyway.

If we support an older version of Node which doesn't have the type that we are using from the `@types/node` package though, then the type won't exist for consumers of the library using that version.

So we should make sure we do two things:

* The major version of `@types/node` should match the **lowest** major version of node we support
* We only use types from the above `@types/node` package (not newer versions, etc)

Reply:
Thank you so much for the detailed explanation. If I have understood correctly, we should use `@types/node` version 10 because the lowest major Node version we support is `10` **, and** we should use only the types defined there. For example, we would have to define our custom type for the `Histogram`, which is not supported on Node 10, but a newer version?
We should also make sure not to let dependabot update the package as I have done lately :cry:.

Reply:
Yes, you've understood correctly. Since version `8.x.x` the major and minor versions in `@types/node` match the major and minor versions in Node. So if we used Node 10.16.3, we need to depend on `@types/node` version `10.16.x` in order for them to match. :slightly_smiling_face:

Reply:
Perfect, thanks!

Reply:
I think I managed to do what you suggested.

One last thing, which maybe is material for a different discussion. After reading [this](https://definitelytyped.org/guides/best-practices.html), I am having second thoughts about the way I changed types definitions (namely, the removal of the `doc` namespace). You implemented them in a way that follows best practices, while my changes don't. As a TS user, do you think this is going to be a problem? Can they remain as they are now? I took a lot of inspiration from [fastify](https://github.com/fastify/fastify/blob/master/fastify.d.ts) types definition when I made those changes.

Reply:
Yes, I implemented them in that way because `@dnlup/doc` exports the main function like:

```js
module.exports = <function>;
```

Which means it needs a "default" in ES Modules. The way that is expressed with Types for TypeScript is via:

```ts
export = Doc;
```

And, the way that a type and function can be combined is via a `namespace` like I did before.

---

You _can leave them as they are now_, but there is one caveat: with the new approach TypeScript will not through a compilation error but a instead a runtime error (so the error won't be discovered until runtime). Compare the following:

### `@dnlup/doc@1.1.0` (namespaces)

```bash
mkdir -p /tmp/doc-1 && cd /tmp/doc-1
yarn add typescript ts-node @dnlup/doc@1.1.0
echo "import Doc from '@dnlup/doc'; console.log(Doc());" > index.ts
./node_modules/.bin/ts-node ./index.ts
```

The above results in:

```
/tmp/doc-1/node_modules/ts-node/src/index.ts:500
    return new TSError(diagnosticText, diagnosticCodes)
           ^
TSError: ⨯ Unable to compile TypeScript:
index.ts:1:8 - error TS1259: Module '"/tmp/doc-1/node_modules/@dnlup/doc/index"' can only be default-imported using the 'esModuleInterop' flag
```

### `@dnlup/doc@2.0.2`

```bash
mkdir -p /tmp/doc-2 && cd /tmp/doc-2
yarn add typescript ts-node @dnlup/doc@2.0.2
echo "import Doc from '@dnlup/doc'; console.log(Doc());" > index.ts
./node_modules/.bin/ts-node ./index.ts
```

The above results in:

```
tmp/doc-2/index.ts:1
import Doc from '@dnlup/doc'; console.log(Doc());
                                             ^
TypeError: doc_1.default is not a function
    at Object.<anonymous> (/tmp/doc-2/index.ts:1:46)
    at Module._compile (internal/modules/cjs/loader.js:1251:30)
    at Module.m._compile (/tmp/doc-2/node_modules/ts-node/src/index.ts:1043:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1272:10)
    at Object.require.extensions.<computed> [as .ts] (/tmp/doc-2/node_modules/ts-node/src/index.ts:1046:12)
    at Module.load (internal/modules/cjs/loader.js:1100:32)
    at Function.Module._load (internal/modules/cjs/loader.js:962:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at main (/tmp/doc-2/node_modules/ts-node/src/bin.ts:225:14)
    at Object.<anonymous> (/tmp/doc-2/node_modules/ts-node/src/bin.ts:512:3
```

---

As you can see, the second approach (without `export = X` and the namespace) results in a runtime error, but the first approach leverages the Type system correctly so that these errors are solved _before_ runtime, at _compilation_ time.

Reply:
Ok, again, thank you for taking the time to write this very detailed explanation. An error caught during compilation is indeed better than one caught at runtime when using TS. It's one of the reasons you might want to use it in the first place. Since this will be landed in a major release, I think we can refactor types definitions to reach that behavior. I'll try in a separate PR, though. Thank you again @acheronfail, your input is precious.